src/main/scala/firrtl/passes/InferReadWrite.scala

```diff
@@ -50,153 +51,137 @@ case class InferReadWriteAnnotation(t: String, tID: TransID)
+            DoPrim(Or, Seq(connects(memPortField(mem, r, "en")),
+                           connects(memPortField(mem, w, "en"))), Nil, BoolType))
+          stmts += Connect(NoInfo, WSubField(rwExp, "addr", ut, FEMALE),
+            Mux(connects(memPortField(mem, w, "en")),
+                connects(memPortField(mem, w, "addr")), 
+                connects(memPortField(mem, r, "addr")), ut))
+        }
       }
-
-    def replaceStmt(s: Statement): Statement =
-      s map replaceStmt map replaceExp match {
-        case Connect(info, loc, exp) if loc == EmptyExpression => EmptyStmt 
-        case s => s
+      readwriters.nonEmpty match {
+        case false => mem
+        case true => mem copy (
+          readers = mem.readers filterNot readers,
+          writers = mem.writers filterNot writers,
+          readwriters = mem.readwriters ++ readwriters)
       }
-    
-    analyze(m.body)
-    Module(m.info, m.name, m.ports, Block((m.body map inferReadWrite map replaceStmt) +: stmts.toSeq))
+    case s => s map inferReadWriteStmt(connects, repl, stmts)
+  }
+
+  def inferReadWrite(m: DefModule) = {
+    val connects = getConnects(m)
+    val repl = new Netlist
+    val stmts = new Statements
+    (m map inferReadWriteStmt(connects, repl, stmts)
+       map replaceStmt(repl)) match {
+      case m: ExtModule => m
+      case m: Module => m copy (body = Block(m.body +: stmts))
+    }
   }
 
-  def run (c:Circuit) = Circuit(c.info, c.modules map {
-    case m: Module => inferReadWrite(m)
-    case m: ExtModule => m
-  }, c.main)
+  def run(c: Circuit) = c copy (modules = c.modules map inferReadWrite)
 }
 
 // Transform input: Middle Firrtl. Called after "HighFirrtlToMidleFirrtl"
 // To use this transform, circuit name should be annotated with its TransId.
-class InferReadWrite(transID: TransID) extends Transform with LazyLogging {
-  def execute(circuit:Circuit, map: AnnotationMap) = 
-    map get transID match {
-      case Some(p) => p get CircuitName(circuit.main) match {
-        case Some(InferReadWriteAnnotation(_, _)) => TransformResult((Seq(
-          InferReadWritePass,
-          CheckInitialization,
-          ResolveKinds,
-          InferTypes,
-          ResolveGenders) foldLeft circuit){ (c, pass) =>
-            val x = Utils.time(pass.name)(pass run c)
-            logger debug x.serialize
-            x
-          }, None, Some(map))
-        case _ => TransformResult(circuit, None, Some(map))
-      }
-      case _ => TransformResult(circuit, None, Some(map))
+class InferReadWrite(transID: TransID) extends Transform with SimpleRun {
+  def passSeq = Seq(
+    InferReadWritePass,
+    CheckInitialization,
+    InferTypes,
+    ResolveKinds,
+    ResolveGenders
+  )
+  def execute(c: Circuit, map: AnnotationMap) = map get transID match {
+    case Some(p) => p get CircuitName(c.main) match {
+      case Some(InferReadWriteAnnotation(_, _)) => run(c, passSeq)
+      case _ => TransformResult(c)
```

Comment:
I think just throw an exception for case _ rather than using TranformResult(c) ?